core/src/main/scala/org/apache/spark/SparkConf.scala

```diff
@@ -373,6 +373,11 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging with Seria
     Option(settings.get(key)).orElse(getDeprecatedConfig(key, this))
   }
 
+  /** Get an optional value, applying variable substitution. */
+  private[spark] def getWithSubstitution(key: String): Option[String] = {
```

Comment:
Hmmm... if I remember the original discussion on the PR that added this code, I chose not to do it to avoid breaking people who might have these references in their configs even though the code wasn't doing substitution.

It might be a good idea to do substitution for all config entries, registered or not, but I'd rather do that in a separate change (and then this private API could be removed).